// tdesign-vue/esm/time-picker/index.js
/**
 * tdesign v0.38.1
 * (c) 2022 tdesign
 * @license MIT
 */

import _TimePicker from './time-picker.js';
import _TimeRangePicker from './time-range-picker.js';
import mapProps from '../utils/map-props.js';
import { withInstall } from '../utils/withInstall.js';
import './style/index.js';
export { EPickerCols, KeyboardDirection } from './interface.js';
import '../_chunks/dep-99305448.js';
import '@babel/runtime/helpers/defineProperty';
import 'dayjs';
import 'lodash/isFunction';
import '../_chunks/dep-ab4703df.js';
import 'vue';
import 'lodash/debounce';
import '../_chunks/dep-e3c17631.js';
import './panel/props.js';
import './props.js';
import './constant.js';
import '../config.js';
import 'tdesign-icons-vue';
import '../utils/mixins.js';
import '../config-provider/config-receiver.js';
import 'lodash/mergeWith';
import '../config-provider/zh_CN_config.js';
import '../config-provider/type.js';
import '../popup/index.js';
import '../popup/popup.js';
import '@babel/runtime/helpers/typeof';
import '@popperjs/core';
import '../utils/classnames.js';
import '../utils/dom.js';
import 'raf';
import 'lodash/isString';
import '../utils/easing.js';
import '../utils/render-tnode.js';
import 'lodash/isObject';
import '../utils/helper.js';
import '@babel/runtime/helpers/objectWithoutProperties';
import '@babel/runtime/helpers/slicedToArray';
import 'lodash/camelCase';
import '../_common/js/utils/set-style.js';
import '../popup/props.js';
import '../popup/container.js';
import '../popup/type.js';
import './panel/index.js';
import '../button/button.js';
import '../loading/index.js';
import '../loading/loading.js';
import '../loading/icon/gradient.js';
import '../_common/js/loading/circle-adapter.js';
import '../_common/js/utils/helper.js';
import '../utils/transfer-dom.js';
import '../loading/props.js';
import '../loading/type.js';
import '../loading/plugin.js';
import '../button/props.js';
import '../utils/ripple.js';
import '../input/index.js';
import '../input/addon.js';
import '../input/input.js';
import '@babel/runtime/helpers/toConsumableArray';
import 'lodash/kebabCase';
import '../utils/event.js';
import '../input/props.js';
import '../input/input-group.js';
import '../input/type.js';
import 'lodash/capitalize';
import './input-items.js';
import 'lodash/isEqual';
import './time-range-picker-props.js';
import './type.js';

var TimePicker = withInstall(mapProps(["value"], {
  model: {
    prop: "value",
    event: "change"
  }
})(_TimePicker));
var TimeRangePicker = withInstall(mapProps(["value"], {
  model: {
    prop: "value",
    event: "change"
  }
})(_TimeRangePicker));

export { TimePicker, TimeRangePicker, TimePicker as default };
