// tdesign-vue/lib/slider/slider-mark.js
/**
 * tdesign v0.38.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var Vue = require('vue');
var config = require('../config.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var Vue__default = /*#__PURE__*/_interopDefaultLegacy(Vue);

var TSliderMark = Vue__default["default"].extend({
  name: "TSliderMark",
  props: {
    mark: {
      type: [Object, Array, String]
    },
    point: {
      type: Number
    }
  },
  methods: {
    changeValue: function changeValue(event) {
      event.stopPropagation();
      this.$emit("change-value", this.point);
    }
  },
  render: function render() {
    var h = arguments[0];
    var label = this.mark;
    return h("div", {
      "class": "".concat(config.prefix, "-slider__mark-text"),
      "on": {
        "click": this.changeValue
      }
    }, [label]);
  }
});

exports["default"] = TSliderMark;
