// tdesign-vue/esm/select/props.js
/**
 * tdesign v0.38.1
 * (c) 2022 tdesign
 * @license MIT
 */

var props = {
  bordered: {
    type: Boolean,
    "default": true
  },
  clearable: Boolean,
  collapsedItems: {
    type: Function
  },
  creatable: Boolean,
  disabled: Boolean,
  empty: {
    type: [String, Function],
    "default": ""
  },
  filter: {
    type: Function
  },
  filterable: Boolean,
  keys: {
    type: Object
  },
  loading: Boolean,
  loadingText: {
    type: [String, Function],
    "default": ""
  },
  max: {
    type: Number,
    "default": 0
  },
  minCollapsedNum: {
    type: Number,
    "default": 0
  },
  multiple: Boolean,
  options: {
    type: Array,
    "default": function _default() {
      return [];
    }
  },
  panelBottomContent: {
    type: [String, Function]
  },
  panelTopContent: {
    type: [String, Function]
  },
  placeholder: {
    type: String,
    "default": void 0
  },
  popupProps: {
    type: Object
  },
  prefixIcon: {
    type: Function
  },
  reserveKeyword: Boolean,
  showArrow: {
    type: Boolean,
    "default": true
  },
  size: {
    type: String,
    "default": "medium",
    validator: function validator(val) {
      return ["small", "medium", "large"].includes(val);
    }
  },
  value: {
    type: [String, Number, Object, Array]
  },
  defaultValue: {
    type: [String, Number, Object, Array]
  },
  valueDisplay: {
    type: Function
  },
  valueType: {
    type: String,
    "default": "value",
    validator: function validator(val) {
      return ["value", "object"].includes(val);
    }
  },
  onBlur: Function,
  onChange: Function,
  onClear: Function,
  onCreate: Function,
  onEnter: Function,
  onFocus: Function,
  onRemove: Function,
  onSearch: Function,
  onVisibleChange: Function
};

export { props as default };
