// tdesign-vue/esm/radio/index.js
/**
 * tdesign v0.38.1
 * (c) 2022 tdesign
 * @license MIT
 */

import _Radio from './radio.js';
import _RadioGroup from './group.js';
import _RadioButton from './radio-button.js';
import mapProps from '../utils/map-props.js';
import { withInstall } from '../utils/withInstall.js';
import './style/index.js';
import './type.js';
import '../_chunks/dep-99305448.js';
import '@babel/runtime/helpers/defineProperty';
import 'vue';
import '../config.js';
import '../utils/classnames.js';
import '../utils/helper.js';
import '@babel/runtime/helpers/objectWithoutProperties';
import '@babel/runtime/helpers/slicedToArray';
import 'lodash/camelCase';
import '../utils/render-tnode.js';
import '@babel/runtime/helpers/typeof';
import 'lodash/isObject';
import './props.js';
import '../utils/event.js';
import 'lodash/isString';
import 'lodash/isNumber';
import './radio-group-props.js';
import 'lodash/capitalize';

var Radio = withInstall(mapProps(["checked"], {
  model: {
    prop: "checked",
    event: "change"
  }
})(_Radio));
var RadioGroup = withInstall(mapProps(["value"], {
  model: {
    prop: "value",
    event: "change"
  }
})(_RadioGroup));
var RadioButton = withInstall(_RadioButton);

export { Radio, RadioButton, RadioGroup, Radio as default };
