// tdesign-vue/es/table/base-table/table-header.js
/**
 * tdesign v0.38.1
 * (c) 2022 tdesign
 * @license MIT
 */

import Vue from 'vue';
import TableCell from './table-cell.js';
import baseTableProps from '../base-table-props.js';
import { prefix } from '../../config.js';
import '../../_chunks/dep-99305448.js';
import '@babel/runtime/helpers/toConsumableArray';
import '@babel/runtime/helpers/defineProperty';
import '../../popup/index.js';
import '../../popup/popup.js';
import '@babel/runtime/helpers/typeof';
import '@popperjs/core';
import '../../utils/classnames.js';
import '../../utils/dom.js';
import 'raf';
import 'lodash/isString';
import '../../utils/easing.js';
import '../../utils/render-tnode.js';
import 'lodash/isObject';
import '../../utils/helper.js';
import '@babel/runtime/helpers/objectWithoutProperties';
import '@babel/runtime/helpers/slicedToArray';
import 'lodash/camelCase';
import '../../_common/js/utils/set-style.js';
import '../../popup/props.js';
import '../../popup/container.js';
import '../../utils/map-props.js';
import '../../utils/withInstall.js';
import 'lodash/capitalize';
import './style/css.js';
import '../../popup/type.js';
import '../util/common.js';
import 'lodash/isFunction';
import '../../utils/event.js';

var TableHeader = Vue.extend({
  name: "".concat(prefix, "-table-header"),
  props: {
    columns: baseTableProps.columns,
    bordered: baseTableProps.bordered
  },
  data: function data() {
    return {
      cacheRowMaxRowspan: []
    };
  },
  methods: {
    renderHeader: function renderHeader() {
      var _this = this;

      var h = this.$createElement;
      var bordered = this.bordered,
          cacheRowMaxRowspan = this.cacheRowMaxRowspan;
      var trPropsList = [];
      this.renderTr(this.columns, 0, trPropsList, true);
      return trPropsList.map(function (thPropsList, rowindex) {
        var currentRowMaxRowspan = cacheRowMaxRowspan[rowindex] || 1;
        return h("tr", [thPropsList.map(function (_ref) {
          var column = _ref.column,
              rowspan = _ref.rowspan,
              colspan = _ref.colspan,
              index = _ref.index,
              hasChildren = _ref.hasChildren,
              isFirstColumn = _ref.isFirstColumn;
          var withBorder = bordered && index === 0 && !isFirstColumn;
          return _this.renderCell(column, hasChildren ? rowspan : currentRowMaxRowspan, colspan, index, withBorder);
        })]);
      });
    },
    renderTr: function renderTr(columns, currentRowIndex, trPropsList, isParentFirstColumn) {
      var _this2 = this,
          _trPropsList$currentR;

      var cacheRowMaxRowspan = this.cacheRowMaxRowspan;
      var currentRowThProps = [];

      if (typeof trPropsList[currentRowIndex] === "undefined") {
        trPropsList[currentRowIndex] = [];
      }

      if (typeof cacheRowMaxRowspan[currentRowIndex] === "undefined") {
        cacheRowMaxRowspan[currentRowIndex] = 1;
      }

      var colspan = 0;
      var rowspan = 1;
      columns.forEach(function (column, index) {
        var children = column.children;

        if (children !== null && children !== void 0 && children.length) {
          var isFirstColumn = isParentFirstColumn && index === 0;

          var _this2$renderTr = _this2.renderTr(children, currentRowIndex + 1, trPropsList, isFirstColumn),
              occupiedCol = _this2$renderTr.colspan,
              occupiedRow = _this2$renderTr.rowspan;

          colspan += occupiedCol;
          rowspan += occupiedRow;
          cacheRowMaxRowspan[currentRowIndex] = Math.max(rowspan, cacheRowMaxRowspan[currentRowIndex]);
          currentRowThProps[index] = {
            rowspan: 1,
            colspan: occupiedCol,
            column: column,
            index: index,
            hasChildren: true,
            isFirstColumn: index === 0 && isParentFirstColumn
          };
        }
      });
      columns.forEach(function (column, index) {
        var children = column.children;

        if (!children || (children === null || children === void 0 ? void 0 : children.length) === 0) {
          currentRowThProps[index] = {
            rowspan: rowspan,
            colspan: 1,
            column: column,
            index: index,
            hasChildren: false,
            isFirstColumn: index === 0 && isParentFirstColumn
          };
          colspan += 1;
        }
      });

      (_trPropsList$currentR = trPropsList[currentRowIndex]).push.apply(_trPropsList$currentR, currentRowThProps);

      return {
        colspan: colspan,
        rowspan: rowspan
      };
    },
    renderCell: function renderCell(column, rowspan, colspan, index, withBorder) {
      var _this3 = this;

      var h = this.$createElement;
      var title = column.title,
          render = column.render,
          scopedSlots = column.scopedSlots;
      var scopedSlotsTitle = scopedSlots === null || scopedSlots === void 0 ? void 0 : scopedSlots.title;
      var customData = {
        type: "title",
        func: "title"
      };
      var customRender;

      if (typeof this.$scopedSlots[scopedSlotsTitle] === "function") {
        customRender = this.$scopedSlots[scopedSlotsTitle];
        console.error("TDesign Table Warn: scopedSlots will be deprecated, please use `cell` and `title` instead.");
      } else if (typeof title === "string") {
        if (typeof this.$scopedSlots[title] === "function") {
          customRender = function customRender(params) {
            return _this3.$scopedSlots[title](params);
          };
        } else {
          customRender = function customRender() {
            return title;
          };
        }
      } else if (typeof title === "function") {
        customRender = function customRender(h, params) {
          return title(h, {
            colIndex: params.colIndex,
            col: params.col
          });
        };
      } else if (typeof render === "function") {
        customRender = render;
        customData.func = "render";
      } else {
        customRender = function customRender() {
          return "";
        };
      }

      var cellData = {
        col: column,
        colIndex: index,
        customData: customData,
        customRender: customRender,
        type: "th",
        withBorder: withBorder
      };
      return h(TableCell, {
        "attrs": {
          "cellData": cellData,
          "colspan": colspan,
          "rowspan": rowspan
        }
      });
    }
  },
  render: function render() {
    var h = arguments[0];
    return h("thead", [this.renderHeader()]);
  }
});

export { TableHeader as default };
