// tdesign-vue/es/date-picker/calendar-presets.js
/**
 * tdesign v0.38.1
 * (c) 2022 tdesign
 * @license MIT
 */

import Vue from 'vue';
import { Button } from '../button/index.js';
import { prefix } from '../config.js';
import '../button/button.js';
import '../_chunks/dep-99305448.js';
import '@babel/runtime/helpers/defineProperty';
import '../utils/classnames.js';
import '../loading/index.js';
import '../loading/loading.js';
import '../loading/icon/gradient.js';
import '../_common/js/loading/circle-adapter.js';
import '../_common/js/utils/set-style.js';
import '../_common/js/utils/helper.js';
import '@babel/runtime/helpers/objectWithoutProperties';
import '@babel/runtime/helpers/slicedToArray';
import '../utils/dom.js';
import 'raf';
import 'lodash/isString';
import '../utils/easing.js';
import '../utils/render-tnode.js';
import '@babel/runtime/helpers/typeof';
import 'lodash/isObject';
import '../utils/transfer-dom.js';
import '../loading/props.js';
import '../utils/withInstall.js';
import 'lodash/capitalize';
import './style/css.js';
import '../loading/type.js';
import '../loading/plugin.js';
import '../button/props.js';
import '../utils/ripple.js';
import '../config-provider/config-receiver.js';
import 'lodash/mergeWith';
import '../config-provider/zh_CN_config.js';
import '../config-provider/type.js';
import '../utils/mixins.js';
import '../button/type.js';

var CalendarPresets = Vue.extend({
  components: {
    TButton: Button
  },
  props: {
    global: {
      type: Object
    },
    presets: {
      type: Object
    },
    onClick: Function
  },
  methods: {
    clickPreset: function clickPreset(value) {
      this.onClick(value);
    }
  },
  render: function render() {
    var _this = this;

    var h = arguments[0];
    var presets = this.presets;
    return h("div", {
      "class": "".concat(prefix, "-date-picker__presets")
    }, [h("ul", [presets && Object.keys(presets).map(function (key) {
      return h("li", {
        "key": key
      }, [h("a", {
        "on": {
          "click": function click() {
            return _this.clickPreset(presets[key]);
          }
        }
      }, [key])]);
    })])]);
  }
});

export { CalendarPresets as default };
