// tdesign-vue/lib/tag/tag.js
/**
 * tdesign v0.38.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _typeof = require('@babel/runtime/helpers/typeof');
var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var tdesignIconsVue = require('tdesign-icons-vue');
var utils_classnames = require('../utils/classnames.js');
var config = require('../config.js');
var tag_props = require('./props.js');
var utils_mixins = require('../utils/mixins.js');
var configProvider_configReceiver = require('../config-provider/config-receiver.js');
var utils_renderTnode = require('../utils/render-tnode.js');
var utils_event = require('../utils/event.js');
require('vue');
require('lodash/mergeWith');
require('../config-provider/zh_CN_config.js');
require('lodash/isObject');
require('../utils/helper.js');
require('@babel/runtime/helpers/objectWithoutProperties');
require('@babel/runtime/helpers/slicedToArray');
require('lodash/camelCase');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _typeof__default = /*#__PURE__*/_interopDefaultLegacy(_typeof);
var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty__default["default"](target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var name = "".concat(config.prefix, "-tag");
var _Tag = utils_mixins["default"](configProvider_configReceiver["default"]("tag")).extend({
  name: "TTag",
  props: _objectSpread({}, tag_props["default"]),
  computed: {
    tagClass: function tagClass() {
      var _ref;

      return ["".concat(name), "".concat(name, "--").concat(this.theme), utils_classnames["default"].SIZE[this.size], "".concat(name, "--").concat(this.variant), this.shape !== "square" && "".concat(name, "--").concat(this.shape), (_ref = {}, _defineProperty__default["default"](_ref, "".concat(name, "--ellipsis"), this.maxWidth), _defineProperty__default["default"](_ref, "".concat(name, "--close"), this.closable), _defineProperty__default["default"](_ref, "".concat(config.prefix, "-is-disabled"), this.disabled), _defineProperty__default["default"](_ref, "".concat(name, "--disabled"), this.disabled), _ref)];
    },
    tagStyle: function tagStyle() {
      if (this.maxWidth) return {
        maxWidth: "".concat(this.maxWidth, "px")
      };
      return {};
    }
  },
  methods: {
    handleClose: function handleClose(e) {
      if (this.disabled) return;
      utils_event.emitEvent(this, "close", {
        e: e
      });
    },
    handleClick: function handleClick(e) {
      if (this.disabled) return;
      utils_event.emitEvent(this, "click", {
        e: e
      });
    },
    getCloseIcon: function getCloseIcon() {
      var _this = this;

      var h = this.$createElement;
      if (!this.closable) return null;
      var iconClassName = "".concat(config.prefix, "-tag__icon-close");

      if (this.global.closeIcon) {
        return this.global.closeIcon(function (component, b) {
          var tProps = _typeof__default["default"](b) === "object" && "attrs" in b ? b.attrs : {};
          return _this.$createElement(component, {
            props: _objectSpread({}, tProps),
            "class": iconClassName,
            nativeOn: {
              click: _this.handleClose
            }
          });
        });
      }

      return h(tdesignIconsVue.CloseIcon, {
        "nativeOn": {
          "click": this.handleClose
        },
        "class": iconClassName
      });
    }
  },
  render: function render() {
    var h = arguments[0];
    var closeIcon = this.getCloseIcon();
    var tagContent = utils_renderTnode.renderContent(this, "default", "content");
    var tagContentWithMaxWidth = h("span", {
      "style": this.tagStyle,
      "class": "".concat(name, "--text")
    }, [tagContent]);
    var icon = utils_renderTnode.renderTNodeJSX(this, "icon");
    return h("span", {
      "class": this.tagClass,
      "on": {
        "click": this.handleClick
      }
    }, [icon, this.maxWidth ? tagContentWithMaxWidth : tagContent, closeIcon]);
  }
});

exports["default"] = _Tag;
