// tdesign-vue/lib/hooks/tnode.js
/**
 * tdesign v0.38.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var VueCompositionAPI = require('@vue/composition-api');
var isEmpty = require('lodash/isEmpty');
var isFunction = require('lodash/isFunction');
var isObject = require('lodash/isObject');
var isString = require('lodash/isString');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var isEmpty__default = /*#__PURE__*/_interopDefaultLegacy(isEmpty);
var isFunction__default = /*#__PURE__*/_interopDefaultLegacy(isFunction);
var isObject__default = /*#__PURE__*/_interopDefaultLegacy(isObject);
var isString__default = /*#__PURE__*/_interopDefaultLegacy(isString);

var isVNode = function isVNode(obj) {
  var vNode = VueCompositionAPI.h("span", "");
  var VNode2 = vNode.constructor;
  return obj instanceof VNode2;
};

function getDefaultNode(options) {
  var defaultNode;

  if (isObject__default["default"](options) && "defaultNode" in options) {
    defaultNode = options.defaultNode;
  } else if (isVNode(options) || isString__default["default"](options)) {
    defaultNode = options;
  }

  return defaultNode;
}
function getParams(options) {
  return isObject__default["default"](options) && "params" in options ? options.params : null;
}
var useTNodeJSX = function useTNodeJSX() {
  var instance = VueCompositionAPI.getCurrentInstance();
  return function (name, options) {
    var params = getParams(options);
    var defaultNode = getDefaultNode(options);
    var slots = instance.setupContext.slots;
    var propsNode;

    if (Object.keys(instance.props).includes(name)) {
      propsNode = instance.props[name];
    }

    if (propsNode === false) return;

    if (slots[name]) {
      return slots[name](params);
    }

    if (isFunction__default["default"](propsNode)) return propsNode(VueCompositionAPI.h, params);

    if (propsNode === true && defaultNode) {
      return defaultNode;
    }

    return propsNode;
  };
};
var useTNodeDefault = function useTNodeDefault() {
  var renderTNodeJSX = useTNodeJSX();
  return function (name, options) {
    var defaultNode = getDefaultNode(options);
    return renderTNodeJSX(name, options) || defaultNode;
  };
};
var useContent = function useContent() {
  var renderTNodeJSX = useTNodeJSX();
  return function (name1, name2, options) {
    var params = getParams(options);
    var defaultNode = getDefaultNode(options);
    var toParams = params ? {
      params: params
    } : void 0;
    var node1 = renderTNodeJSX(name1, toParams);
    var node2 = renderTNodeJSX(name2, toParams);
    var res = isEmpty__default["default"](node1) ? node2 : node1;
    return isEmpty__default["default"](res) ? defaultNode : res;
  };
};

exports.getDefaultNode = getDefaultNode;
exports.getParams = getParams;
exports.useContent = useContent;
exports.useTNodeDefault = useTNodeDefault;
exports.useTNodeJSX = useTNodeJSX;
