// tdesign-vue/lib/select-input/useSingle.js
/**
 * tdesign v0.38.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var helper = require('../_chunks/dep-c2bd70fb.js');
var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var VueCompositionAPI = require('@vue/composition-api');
var isObject = require('lodash/isObject');
var pick = require('lodash/pick');
var input_index = require('../input/index.js');
var loading_index = require('../loading/index.js');
var hooks_tnode = require('../hooks/tnode.js');
var hooks_useDefaultValue = require('../hooks/useDefaultValue.js');
var configProvider_useConfig = require('../config-provider/useConfig.js');
require('../input/addon.js');
require('vue');
require('../config.js');
require('../input/input.js');
require('@babel/runtime/helpers/toConsumableArray');
require('@babel/runtime/helpers/typeof');
require('tdesign-icons-vue');
require('lodash/camelCase');
require('lodash/kebabCase');
require('../utils/helper.js');
require('@babel/runtime/helpers/objectWithoutProperties');
require('../config-provider/config-receiver.js');
require('lodash/mergeWith');
require('../config-provider/zh_CN_config.js');
require('../utils/mixins.js');
require('../utils/classnames.js');
require('../utils/event.js');
require('../input/props.js');
require('../utils/render-tnode.js');
require('../input/input-group.js');
require('../utils/map-props.js');
require('../utils/withInstall.js');
require('lodash/capitalize');
require('../loading/loading.js');
require('../loading/icon/gradient.js');
require('../_common/js/loading/circle-adapter.js');
require('../_common/js/utils/set-style.js');
require('../_common/js/utils/helper.js');
require('../utils/dom.js');
require('raf');
require('lodash/isString');
require('../utils/easing.js');
require('../utils/transfer-dom.js');
require('../loading/props.js');
require('../loading/plugin.js');
require('lodash/isEmpty');
require('lodash/isFunction');
require('lodash/cloneDeep');
require('../config-provider/context.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);
var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);
var isObject__default = /*#__PURE__*/_interopDefaultLegacy(isObject);
var pick__default = /*#__PURE__*/_interopDefaultLegacy(pick);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty__default["default"](target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var COMMON_PROPERTIES = ["status", "clearable", "disabled", "label", "placeholder", "readonly", "suffix", "suffixIcon", "onPaste", "onEnter", "onMouseenter", "onMouseleave"];
var DEFAULT_KEYS = {
  label: "label",
  value: "value"
};

function getInputValue(value, keys) {
  var iKeys = keys || DEFAULT_KEYS;
  return isObject__default["default"](value) ? value[iKeys.label] : value;
}

function useSingle(props, context) {
  var _props$defaultInputVa;

  var instance = VueCompositionAPI.getCurrentInstance();

  var _toRefs = VueCompositionAPI.toRefs(props),
      value = _toRefs.value,
      keys = _toRefs.keys,
      propsInputValue = _toRefs.inputValue;

  var classPrefix = configProvider_useConfig.usePrefixClass();

  var _useDefaultValue = hooks_useDefaultValue["default"](propsInputValue, (_props$defaultInputVa = props.defaultInputValue) !== null && _props$defaultInputVa !== void 0 ? _props$defaultInputVa : "", props.onInputChange, "inputValue", "input-change"),
      _useDefaultValue2 = _slicedToArray__default["default"](_useDefaultValue, 2),
      inputValue = _useDefaultValue2[0],
      setInputValue = _useDefaultValue2[1];

  var inputRef = VueCompositionAPI.ref();
  var renderTNode = hooks_tnode.useTNodeJSX();
  var commonInputProps = VueCompositionAPI.computed(function () {
    return pick__default["default"](props, COMMON_PROPERTIES);
  });

  var onInnerClear = function onInnerClear(context2) {
    var _context2$e, _props$onClear;

    context2 === null || context2 === void 0 ? void 0 : (_context2$e = context2.e) === null || _context2$e === void 0 ? void 0 : _context2$e.stopPropagation();
    (_props$onClear = props.onClear) === null || _props$onClear === void 0 ? void 0 : _props$onClear.call(props, context2);
    instance.emit("clear", context2);
    setInputValue("", {
      trigger: "clear"
    });
  };

  var onInnerInputChange = function onInnerInputChange(value2, context2) {
    if (props.allowInput) {
      setInputValue(value2, _objectSpread(_objectSpread({}, context2), {}, {
        trigger: "input"
      }));
    }
  };

  var renderSelectSingle = function renderSelectSingle(h, popupVisible) {
    var singleValueDisplay = renderTNode("valueDisplay");
    var displayedValue = popupVisible && props.allowInput ? inputValue.value : getInputValue(value.value, keys.value);
    var prefixContent = [singleValueDisplay, renderTNode("label")];

    var inputProps = _objectSpread(_objectSpread(_objectSpread({}, commonInputProps.value), props.inputProps), {}, {
      value: singleValueDisplay ? void 0 : displayedValue,
      label: prefixContent.length ? function () {
        return prefixContent;
      } : void 0,
      autoWidth: props.autoWidth,
      readonly: !props.allowInput,
      placeholder: singleValueDisplay ? "" : props.placeholder,
      suffixIcon: !props.disabled && props.loading ? function () {
        return h(loading_index.Loading, {
          "attrs": {
            "loading": true,
            "size": "small"
          }
        });
      } : props.suffixIcon,
      showClearIconOnEmpty: Boolean(props.clearable && inputValue.value),
      inputClass: _defineProperty__default["default"]({}, "".concat(classPrefix.value, "-input--focused"), popupVisible)
    });

    return h(input_index.Input, helper.helper([{
      "ref": "inputRef"
    }, {
      "props": inputProps
    }, {
      "scopedSlots": context.slots,
      "on": {
        "change": onInnerInputChange,
        "clear": onInnerClear,
        "blur": function blur(val, context2) {
          var _props$onBlur;

          (_props$onBlur = props.onBlur) === null || _props$onBlur === void 0 ? void 0 : _props$onBlur.call(props, value.value, _objectSpread(_objectSpread({}, context2), {}, {
            inputValue: val
          }));
          instance.emit("blur", value.value, _objectSpread(_objectSpread({}, context2), {}, {
            inputValue: val
          }));
        },
        "enter": function enter(val, context2) {
          var _props$onEnter;

          (_props$onEnter = props.onEnter) === null || _props$onEnter === void 0 ? void 0 : _props$onEnter.call(props, value.value, _objectSpread(_objectSpread({}, context2), {}, {
            inputValue: val
          }));
          instance.emit("enter", value.value, _objectSpread(_objectSpread({}, context2), {}, {
            inputValue: val
          }));
        },
        "focus": function focus(val, context2) {
          var _props$onFocus;

          (_props$onFocus = props.onFocus) === null || _props$onFocus === void 0 ? void 0 : _props$onFocus.call(props, value.value, _objectSpread(_objectSpread({}, context2), {}, {
            inputValue: val
          }));
          instance.emit("focus", value.value, _objectSpread(_objectSpread({}, context2), {}, {
            tagInputValue: val
          }));
          !popupVisible && setInputValue(getInputValue(value.value, keys.value), _objectSpread(_objectSpread({}, context2), {}, {
            trigger: "input"
          }));
        }
      }
    }]));
  };

  return {
    inputRef: inputRef,
    commonInputProps: commonInputProps,
    onInnerClear: onInnerClear,
    renderSelectSingle: renderSelectSingle
  };
}

exports["default"] = useSingle;
