// tdesign-vue/lib/form/const.js
/**
 * tdesign v0.38.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var config = require('../config.js');

var FORM_ITEM_CLASS_PREFIX = "".concat(config.prefix, "-form-item__");
var form = "".concat(config.prefix, "-form");
var input = "".concat(config.prefix, "-input");
var is = "".concat(config.prefix, "-is");
var CLASS_NAMES = {
  form: form,
  label: "".concat(form, "__label"),
  labelTop: "".concat(form, "__label--top"),
  inline: "".concat(form, "-inline"),
  formItem: "".concat(form, "__item"),
  formItemWithHelp: "".concat(form, "__item-with-help"),
  formItemWithExtra: "".concat(form, "__item-with-extra"),
  controls: "".concat(form, "__controls"),
  controlsContent: "".concat(form, "__controls-content"),
  status: "".concat(form, "__status"),
  help: "".concat(form, "__help"),
  extra: "".concat(input, "__extra"),
  success: "".concat(is, "-success"),
  successBorder: "".concat(form, "--success-border"),
  error: "".concat(is, "-error"),
  warning: "".concat(is, "-warning")
};
var FORM_CONTROLE_COMPONENTS = ["TInput", "TInputNumber", "TTextarea", "TCascader", "TSelect", "TOption", "TSwitch", "TCheckbox", "TCheckboxGroup", "TRadio", "TRadioGroup", "TTreeSelect", "TDatePicker", "TTimePicker", "TUpload", "TTransfer", "TSlider"];

exports.CLASS_NAMES = CLASS_NAMES;
exports.FORM_CONTROLE_COMPONENTS = FORM_CONTROLE_COMPONENTS;
exports.FORM_ITEM_CLASS_PREFIX = FORM_ITEM_CLASS_PREFIX;
