// tdesign-vue/esm/layout/footer.js
/**
 * tdesign v0.38.1
 * (c) 2022 tdesign
 * @license MIT
 */

import _defineProperty from '@babel/runtime/helpers/defineProperty';
import Vue from 'vue';
import { prefix } from '../config.js';
import props from './footer-props.js';
import { renderTNodeJSX } from '../utils/render-tnode.js';
import '@babel/runtime/helpers/typeof';
import 'lodash/isObject';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var _Footer = Vue.extend({
  name: "TFooter",
  props: _objectSpread({}, props),
  methods: {
    renderContent: function renderContent() {
      return this.$scopedSlots["default"] ? this.$scopedSlots["default"](null) : "";
    }
  },
  render: function render() {
    var h = arguments[0];
    var styles = this.height ? {
      height: this.height
    } : {};
    return h("footer", {
      "class": "".concat(prefix, "-layout__footer"),
      "style": styles
    }, [renderTNodeJSX(this, "default")]);
  }
});

export { _Footer as default };
