// tdesign-vue/lib/tag-input/tag-input.js
/**
 * tdesign v0.38.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var helper = require('../_chunks/dep-c2bd70fb.js');
var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var VueCompositionAPI = require('@vue/composition-api');
var tdesignIconsVue = require('tdesign-icons-vue');
var input_index = require('../input/index.js');
var tagInput_props = require('./props.js');
var config = require('../config.js');
var utils_renderTnode = require('../utils/render-tnode.js');
var tagInput_useTagScroll = require('./useTagScroll.js');
var tagInput_useTagList = require('./useTagList.js');
var tagInput_useHover = require('./useHover.js');
var hooks_useDefaultValue = require('../hooks/useDefaultValue.js');
require('../input/addon.js');
require('vue');
require('../input/input.js');
require('@babel/runtime/helpers/toConsumableArray');
require('@babel/runtime/helpers/typeof');
require('lodash/camelCase');
require('lodash/kebabCase');
require('../utils/helper.js');
require('@babel/runtime/helpers/objectWithoutProperties');
require('../config-provider/config-receiver.js');
require('lodash/mergeWith');
require('../config-provider/zh_CN_config.js');
require('../utils/mixins.js');
require('../utils/classnames.js');
require('../utils/event.js');
require('../input/props.js');
require('lodash/isObject');
require('../input/input-group.js');
require('../utils/map-props.js');
require('../utils/withInstall.js');
require('lodash/capitalize');
require('../tag/index.js');
require('../tag/tag.js');
require('../tag/props.js');
require('../tag/check-tag.js');
require('../tag/check-tag-props.js');
require('../hooks/useVModel.js');
require('../hooks/tnode.js');
require('lodash/isEmpty');
require('lodash/isFunction');
require('lodash/isString');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);
var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty__default["default"](target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var NAME_CLASS = "".concat(config.prefix, "-tag-input");
var CLEAR_CLASS = "".concat(config.prefix, "-tag-input__suffix-clear");
var BREAK_LINE_CLASS = "".concat(config.prefix, "-tag-input--break-line");
var _TagInput = VueCompositionAPI.defineComponent({
  name: "TTagInput",
  props: _objectSpread({}, tagInput_props["default"]),
  setup: function setup(props2, context) {
    var _toRefs = VueCompositionAPI.toRefs(props2),
        inputValue = _toRefs.inputValue;

    var _useDefaultValue = hooks_useDefaultValue["default"](inputValue, props2.defaultInputValue, props2.onInputChange, "inputValue", "input-change"),
        _useDefaultValue2 = _slicedToArray__default["default"](_useDefaultValue, 2),
        tInputValue = _useDefaultValue2[0],
        setTInputValue = _useDefaultValue2[1];

    var _toRefs2 = VueCompositionAPI.toRefs(props2),
        excessTagsDisplayType = _toRefs2.excessTagsDisplayType,
        readonly = _toRefs2.readonly,
        disabled = _toRefs2.disabled,
        clearable = _toRefs2.clearable,
        placeholder = _toRefs2.placeholder;

    var _useHover = tagInput_useHover["default"]({
      readonly: props2.readonly,
      disabled: props2.disabled,
      onMouseenter: props2.onMouseenter,
      onMouseleave: props2.onMouseleave
    }),
        isHover = _useHover.isHover,
        addHover = _useHover.addHover,
        cancelHover = _useHover.cancelHover;

    var _useTagScroll = tagInput_useTagScroll["default"](props2),
        scrollToRight = _useTagScroll.scrollToRight,
        onWheel = _useTagScroll.onWheel,
        scrollToRightOnEnter = _useTagScroll.scrollToRightOnEnter,
        scrollToLeftOnLeave = _useTagScroll.scrollToLeftOnLeave,
        tagInputRef = _useTagScroll.tagInputRef;

    var _useTagList = tagInput_useTagList["default"](props2),
        tagValue = _useTagList.tagValue,
        onInnerEnter = _useTagList.onInnerEnter,
        onInputBackspaceKeyUp = _useTagList.onInputBackspaceKeyUp,
        clearAll = _useTagList.clearAll,
        renderLabel = _useTagList.renderLabel,
        onClose = _useTagList.onClose;

    var classes = VueCompositionAPI.computed(function () {
      return [NAME_CLASS, _defineProperty__default["default"]({}, BREAK_LINE_CLASS, excessTagsDisplayType.value === "break-line")];
    });
    var tagInputPlaceholder = VueCompositionAPI.computed(function () {
      var _tagValue$value;

      return !((_tagValue$value = tagValue.value) !== null && _tagValue$value !== void 0 && _tagValue$value.length) ? placeholder.value : "";
    });
    var showClearIcon = VueCompositionAPI.computed(function () {
      var _tagValue$value2;

      return Boolean(!readonly.value && !disabled.value && clearable.value && isHover.value && (((_tagValue$value2 = tagValue.value) === null || _tagValue$value2 === void 0 ? void 0 : _tagValue$value2.length) || tInputValue.value));
    });

    var onInputEnter = function onInputEnter(value, context2) {
      setTInputValue("", {
        e: context2.e,
        trigger: "enter"
      });
      onInnerEnter(value, context2);
      VueCompositionAPI.nextTick(function () {
        scrollToRight();
      });
    };

    var onClick = function onClick() {
      tagInputRef.value.focus();
    };

    var onClearClick = function onClearClick(ctx) {
      var _props2$onClear;

      clearAll(ctx);
      setTInputValue("", {
        e: ctx.e,
        trigger: "clear"
      });
      (_props2$onClear = props2.onClear) === null || _props2$onClear === void 0 ? void 0 : _props2$onClear.call(props2, ctx);
      context.emit("clear", ctx);
    };

    return {
      tagValue: tagValue,
      tInputValue: tInputValue,
      isHover: isHover,
      tagInputPlaceholder: tagInputPlaceholder,
      showClearIcon: showClearIcon,
      tagInputRef: tagInputRef,
      setTInputValue: setTInputValue,
      addHover: addHover,
      cancelHover: cancelHover,
      onInputEnter: onInputEnter,
      onInnerEnter: onInnerEnter,
      onInputBackspaceKeyUp: onInputBackspaceKeyUp,
      renderLabel: renderLabel,
      onWheel: onWheel,
      scrollToRightOnEnter: scrollToRightOnEnter,
      scrollToLeftOnLeave: scrollToLeftOnLeave,
      onClick: onClick,
      onClearClick: onClearClick,
      onClose: onClose,
      classes: classes
    };
  },
  render: function render(h) {
    var _this = this;

    var suffixIconNode = this.showClearIcon ? h(tdesignIconsVue.CloseCircleFilledIcon, {
      "class": CLEAR_CLASS,
      "on": {
        "click": this.onClearClick
      }
    }) : utils_renderTnode.renderTNodeJSX(this, "suffixIcon");
    var displayNode = utils_renderTnode.renderTNodeJSX(this, "valueDisplay", {
      params: {
        value: this.tagValue,
        onClose: function onClose(index, item) {
          return _this.onClose({
            index: index,
            item: item
          });
        }
      }
    });

    var _label = utils_renderTnode.renderTNodeJSX(this, "label", {
      silent: true
    });

    return h(input_index.Input, helper.helper([{
      "ref": "tagInputRef",
      "attrs": {
        "readonly": this.readonly
      }
    }, this.inputProps, {
      "attrs": {
        "value": this.tInputValue,
        "autoWidth": this.autoWidth,
        "size": this.size,
        "disabled": this.disabled,
        "label": function label() {
          return _this.renderLabel({
            displayNode: displayNode,
            label: _label
          }, h);
        },
        "tips": this.tips,
        "status": this.status,
        "placeholder": this.tagInputPlaceholder,
        "suffix": this.suffix,
        "suffixIcon": function suffixIcon() {
          return suffixIconNode;
        }
      },
      "on": {
        "change": function change(val, context) {
          _this.setTInputValue(val, _objectSpread(_objectSpread({}, context), {}, {
            trigger: "input"
          }));
        },
        "mousewheel": this.onWheel,
        "enter": this.onInputEnter,
        "keyup": this.onInputBackspaceKeyUp,
        "mouseenter": function mouseenter(context) {
          _this.addHover(context);

          _this.scrollToRightOnEnter();
        },
        "mouseleave": function mouseleave(context) {
          _this.cancelHover(context);

          _this.scrollToLeftOnLeave();
        },
        "focus": function focus(inputValue, context) {
          var _this$onFocus;

          (_this$onFocus = _this.onFocus) === null || _this$onFocus === void 0 ? void 0 : _this$onFocus.call(_this, _this.tagValue, {
            e: context.e,
            inputValue: inputValue
          });

          _this.$emit("focus", _this.tagValue, {
            e: context.e,
            inputValue: inputValue
          });
        },
        "blur": function blur(inputValue, context) {
          var _this$onBlur;

          (_this$onBlur = _this.onBlur) === null || _this$onBlur === void 0 ? void 0 : _this$onBlur.call(_this, _this.tagValue, {
            e: context.e,
            inputValue: inputValue
          });

          _this.$emit("blur", _this.tagValue, {
            e: context.e,
            inputValue: inputValue
          });
        },
        "paste": function paste(context) {
          var _this$onPaste;

          (_this$onPaste = _this.onPaste) === null || _this$onPaste === void 0 ? void 0 : _this$onPaste.call(_this, context);

          _this.$emit("paste", context);
        }
      },
      "class": this.classes
    }]));
  }
});

exports["default"] = _TagInput;
