// tdesign-vue/lib/layout/aside.js
/**
 * tdesign v0.38.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var Vue = require('vue');
var config = require('../config.js');
var layout_asideProps = require('./aside-props.js');
var utils_renderTnode = require('../utils/render-tnode.js');
require('@babel/runtime/helpers/typeof');
require('lodash/isObject');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);
var Vue__default = /*#__PURE__*/_interopDefaultLegacy(Vue);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty__default["default"](target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var _Aside = Vue__default["default"].extend({
  name: "TAside",
  props: _objectSpread({}, layout_asideProps["default"]),
  data: function data() {
    return {};
  },
  inject: {
    layout: {
      "default": void 0
    }
  },
  mounted: function mounted() {
    this.layout.hasSider = true;
  },
  destroyed: function destroyed() {
    this.layout.hasSider = false;
  },
  render: function render() {
    var h = arguments[0];
    var styles = this.width ? {
      width: this.width
    } : {};
    return h("aside", {
      "class": "".concat(config.prefix, "-layout__sider"),
      "style": styles
    }, [utils_renderTnode.renderTNodeJSX(this, "default")]);
  }
});

exports["default"] = _Aside;
