// tdesign-vue/es/layout/layout.js
/**
 * tdesign v0.38.1
 * (c) 2022 tdesign
 * @license MIT
 */

import _defineProperty from '@babel/runtime/helpers/defineProperty';
import Vue from 'vue';
import { prefix } from '../config.js';
import { renderTNodeJSX } from '../utils/render-tnode.js';
import '@babel/runtime/helpers/typeof';
import 'lodash/isObject';

var name = "".concat(prefix, "-layout");
var _Layout = Vue.extend({
  name: "TLayout",
  data: function data() {
    return {
      hasSider: false
    };
  },
  provide: function provide() {
    return {
      layout: this
    };
  },
  methods: {
    renderContent: function renderContent() {
      return this.$scopedSlots["default"] ? this.$scopedSlots["default"](null) : "";
    }
  },
  computed: {
    classes: function classes() {
      return [name, _defineProperty({}, "".concat(name, "--with-sider"), this.hasSider)];
    }
  },
  render: function render() {
    var h = arguments[0];
    return h("section", {
      "class": this.classes
    }, [renderTNodeJSX(this, "default")]);
  }
});

export { _Layout as default };
